File: gfg/2024_7_july/python/01_make_binary_tree_from_linked_list.py
# CODE

class Tree:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

def convert(head):
    nodes = [] # for storing tree nodes
    while head:
        nodes.append(Tree(head.data))
        head = head.next
    n = len(nodes)
    for i, node in enumerate(nodes):
        if 2 * i + 1 < n:
            nodes[i].left = nodes[2 * i + 1]
        if 2 * i + 2 < n:
            nodes[i].right = nodes[2 * i + 2]
    return nodes[0]

File: gfg/2024_7_july/c++/01_make_binary_tree_from_linked_list.cpp
// TIME COMPLEXITY
// O(n)

// SPACE COMPLEXITY
// O(n)

// CODE

void convert(Node *head, TreeNode *&root) {
    vector<TreeNode*> tree;
    Node *temp = head;
    while(temp != NULL){
        tree.push_back(new TreeNode(temp->data));
        temp = temp -> next;
    }
    for(int i = 0; i < tree.size(); i++){
        int left = 2*i + 1;
        int right = 2*i + 2;
        if(left < tree.size()){
            tree[i] -> left = tree[left];
        }
        if(right < tree.size()){
            tree[i] -> right = tree[right];
        }
    }
    root = tree[0];
}